Civi/Core/Paths.php

```diff
@@ -56,6 +56,21 @@ public function __construct() {
       ->register('civicrm.files', function () {
         return \CRM_Core_Config::singleton()->userSystem->getDefaultFileStorage();
       })
+      ->register('civicrm.private', function () {
+        return [
+          'path' => \CRM_Utils_File::baseFilePath(),
+        ];
+      })
+      ->register('civicrm.log', function () {
+        return [
+          'path' => \Civi::paths()->getPath('[civicrm.private]/ConfigAndLog'),
+        ];
+      })
+      ->register('civicrm.compile', function () {
+        return [
+          'path' => \Civi::paths()->getPath('[civicrm.private]/templates_c'),
```

Comment:
Is this a problem for `$file = CIVICRM_TEMPLATE_COMPILEDIR . "/CachedCiviContainer.{$envId}.php";` in \Civi\Core\Container? Can that file be in a private directory? I'm guessing 

Same with `CRM_Extension_ClassLoader::getCacheFile()` there's this line:
`$file = CIVICRM_TEMPLATE_COMPILEDIR . "/CachedExtLoader.{$envId}.php";`

Along with this maybe the only place that references `CIVICRM_TEMPLATE_COMPILEDIR` should be `\CRM_Utils_File::baseFilePath` so that people can't set the template directory twice. The other references could call `[civicrm.private]/templates_c`?

Reply:
Oops. I guess `\Civi\Core\Container::loadContainer()` can't expect `Civi\Core\Paths` to be available. So `CIVICRM_TEMPLATE_COMPILEDIR` needs to stick around at least for this.

On Pantheon I've been setting `CIVICRM_TEMPLATE_COMPILEDIR` to the tmp directory to deal with a special issue there, and that works so long as `[civicrm.files]` is also set to point to the regular file system.

Reply:
This is how civicrm.settings.php would be set up to work on Pantheon:

```
if (!defined('CIVICRM_TEMPLATE_COMPILEDIR')) {
  define( 'CIVICRM_TEMPLATE_COMPILEDIR', '/tmp/civicrm/templates_c/');
}
$civicrm_paths['civicrm.compile']['path'] = CIVICRM_TEMPLATE_COMPILEDIR;
$civicrm_paths['civicrm.files']['path'] = '/app/sites/default/files/civicrm/';
$civicrm_paths['civicrm.private']['path'] = '/app/sites/default/files/private/civicrm/';
```

For now might not need `[civicrm.compile]` since it's not being called anywhere, I think.

Reply:
So maybe the only thing I'd borrow from https://github.com/civicrm/civicrm-core/pull/14719 is to make this one:

```
->register('civicrm.compile', function () {
        return [
          'path' => CIVICRM_TEMPLATE_COMPILEDIR,
        ];
      })
```

The other paths are probably fine in this PR.

Reply:
I *think* most remaining references to `CIVICRM_TEMPLATE_COMPILEDIR` come late enough during bootstrap that `Civi::paths()->getPath()` is viable substitute... but I'm gonna try it with some funny-cases (like `GenCode` and `install/index.php`) to be sure.

Reply:
@herbdool So, going along the lines of your first comment, I've pushed an updated branch which patches `\Civi\Core\Container::loadContainer()`, `CRM_Extension_ClassLoader`, and `CRM_Core_IDS` to replace `CIVICRM_TEMPLATE_COMPILEDIR` with `Civi::paths()->getFoo()`. 

This brings it down to only *one* reference to `CIVICRM_TEMPLATE_COMPILEDIR` (in `CRM_Utils_File::baseFilePath()`). 

I haven't tested the following claim, but it *should* mean that it is valid to completely omit `CIVICRM_TEMPLATE_COMPILEDIR` and instead specify `civicrm.files` and `civicrm.private`.  By setting those two variables, one bypasses any remaining code-paths that rely directly on `baseFilePath()` or `CIVICRM_TEMPLATE_COMPILEDIR`.

Reply:
Along the lines of your last comment, I updated the `civicrm.compile` calculation so that it consults `CIVICRM_TEMPLATE_COMPILEDIR`. Formally, the effective value of `civicrm.compile` will be the first defined item from this list:

1. `$civicrm_paths['civicrm.compile']['path']`
2. `CIVICRM_TEMPLATE_COMPILEDIR`
3. `Civi::paths()->getPath('[civicrm.private]/templates_c')`

I *think* this means that we get backward-compatibility for existing deployments as well as supporting a sensible style for new deployments:

* (*Backwards compatible style*) If you have an existing deployment and existing `civicrm.settings.php` (with `define('CIVICRM_TEMPLATE_COMPILEDIR'...)` and no customization to `$civicrm_paths`, etc) - then `CIVICRM_TEMPLATE_COMPILEDIR` is still used.
* (*New style*) If you make a new/revised deployment with a new/revised `civicrm.settings.php` (omitting `define('CIVICRM_TEMPLATE_COMPILEDIR', ...)`; instead, using `$civicrm_paths`), then:
    * The default value of `civicrm.compile` is `[civicrm.private]/templates_c`
    * The default value of `civicrm.log` is `[civicrm.private]/ConfigAndLog`
    * Either can be customized directly (without impacting the parent).